test/integration/original_ip_detection_integration_test.cc

```diff
@@ -0,0 +1,75 @@
+#include "test/integration/http_integration.h"
+#include "test/test_common/registry.h"
+#include "test/test_common/utility.h"
+
+#include "gtest/gtest.h"
+
+using testing::HasSubstr;
+
+namespace Envoy {
+namespace Formatter {
+
+class OriginalIPDetectionIntegrationTest : public testing::Test, public HttpIntegrationTest {
+public:
+  OriginalIPDetectionIntegrationTest()
+      : HttpIntegrationTest(Http::CodecClient::Type::HTTP1, Network::Address::IpVersion::v4) {}
+};
+
+class CustomHeaderDetection : public Http::OriginalIPDetection {
+public:
+  CustomHeaderDetection(const std::string& header_name) : header_name_(header_name) {}
+
+  Network::Address::InstanceConstSharedPtr
+  detect(struct Http::OriginalIPDetectionParams& params) override {
+    auto hdr = params.request_headers.get(Http::LowerCaseString(header_name_));
+    if (hdr.empty()) {
+      return nullptr;
+    }
+    auto header_value = hdr[0]->value().getStringView();
+    return std::make_shared<Network::Address::Ipv4Instance>(std::string(header_value));
```

Comment:
I actually thought about adding an extension that just reads the real remote IP from a header.. I guess since we'll be adding the XFF code as an extension, I might as well add `CustomHeaderDetection` as an extension as well.